File: 2020/day01/main.cpp
#include <cassert>
#include <cstdint>
#include <iostream>
#include <vector>
#include <set>
#include <map>

namespace {

std::int64_t Solve01(const std::vector<std::int64_t> inputs) {
    std::set<std::int64_t> s;
    for (auto input : inputs) {
        s.insert(input);
    }

    for (auto input : inputs) {
        auto diff = 2020 - input;
        if (s.find(diff) != s.end()) {
            return input * diff;
        }
    }

    assert(!"never reach here");
    return 0;
}

std::int64_t Solve02(const std::vector<std::int64_t> inputs) {
    std::map<std::int64_t, std::vector<std::int64_t>> m;
    for (size_t i = 0; i < inputs.size() - 1; ++i) {
        for (size_t j = i + 1; j < inputs.size(); ++j) {
            auto sum = inputs[i] + inputs[j];
            m[sum] = std::vector<std::int64_t>{inputs[i], inputs[j]};
        }
    }

    for (auto input : inputs) {
        auto diff = 2020 - input;
        if (m.find(diff) != m.end()) {
            return input * m[diff][0] * m[diff][1];
        }
    }

    assert(!"never reach here");
    return 0;
}

void Test01() {
    std::vector<std::int64_t> inputs{
        1721, 979, 366, 299, 675, 1456,
    };

    assert(Solve01(inputs) == 514579);
}

void Test02() {
    std::vector<std::int64_t> inputs{
        1721, 979, 366, 299, 675, 1456,
    };

    assert(Solve02(inputs) == 241861950);
}

} // namespace

int main() {
    Test01();
    Test02();

    std::vector<std::int64_t> inputs;
    std::int64_t val;

    while (std::cin >> val) {
        inputs.push_back(val);
    }

    auto answer01 = Solve01(inputs);
    assert(answer01 == 157059);
    std::cout << "Part01: " << answer01 << std::endl;

    auto answer02 = Solve02(inputs);
    assert(answer02 == 165080960);
    std::cout << "Part02: " << answer02 << std::endl;
    return 0;
}

File: 2020/day01/rust/src/main.rs
use std::collections::{HashMap, HashSet};
use std::io::{self, Read};

fn solve01(nums: &Vec<i32>) -> i32 {
    let hs: HashSet<&i32> = nums.iter().collect();
    for num in nums.iter() {
        let diff = 2020 - num;
        if hs.contains(&diff) {
            return num * diff;
        }
    }

    panic!("never reash here");
}

fn solve02(nums: &Vec<i32>) -> i32 {
    let mut m: HashMap<i32, i32> = HashMap::new();
    for (i, v1) in nums.iter().enumerate() {
        for v2 in nums.iter().skip(i + 1) {
            m.insert(v1 + v2, v1 * v2);
        }
    }

    for v in nums.iter() {
        let diff = 2020 - v;
        if let Some(n) = m.get(&diff) {
            return v * n;
        }
    }

    panic!("never reash here");
}

fn parse_input(input: &str) -> Vec<i32> {
    input.lines().filter_map(|s| s.parse().ok()).collect()
}

fn main() -> io::Result<()> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    let nums = parse_input(&input);

    println!("Part01: {}", solve01(&nums));
    println!("Part02: {}", solve02(&nums));

    Ok(())
}

#[test]
fn test() {
    let input = r#"1721
979
366
299
675
456"#;

    let nums = parse_input(input);
    assert_eq!(nums, vec![1721, 979, 366, 299, 675, 456]);
    assert_eq!(solve01(&nums), 514579);
    assert_eq!(solve02(&nums), 241_861_950);
}
